storage/storage_dest.go

```diff
@@ -604,13 +636,38 @@ func (s *storageImageDestination) commitLayer(index int, info addedLayerInfo, si
 	diffOutput, ok := s.diffOutputs[info.digest]
 	s.lock.Unlock()
 	if ok {
+		if s.manifest == nil {
+			logrus.Debugf("Skipping commit for TOC=%q, manifest not yet available")
+			return nil
```

Comment:
(Just looking at the recent changes)

I can’t see how just silently indicating success without doing anything else can work here. The push code is then going to call `commitLayer` for the following layer, and incorrectly set `lastLayer` to `""`.

Reply:
from what I could see, `commitLayer` is called twice.   When it is called from `queueOrCommit` the `s.manifest` is nil so we avoid committing the layer at this stage.

Then it is called again from `Commit` and this time `s.manifest` is present and the layer is created.

With partial images, there is no much gain from the parallel `queueOrCommit` optimization as anyway the commit phase is just renaming a directory so it is fine to do it sequentially from `Commit`

Reply:
Yes, but it’s the _first_ `commitLayer` of a child of the layer in question that is, I think, going to create problems (a layer with no parent instead of the right parent), and a second commit of the child is then going to be a no-op.

At a glance, it seems to me that `commitLayer` needs to be able to say “not yet, sorry”, causing the `queueOrCommit` state machine to just give up and wait for the final `Commit` to do all the work. Alternatively, `private.ImageDestination` can be extended so that the caller provides the config blob early to transports that ask for it.

(… but I’m not sure this is on the critical path for now. I am still deeply worried about how `TryReusingBlobWithOptions` finds a TOC-based match, in the first place, and I don’t remember, any more, all of the details of how that works / interacts with this DiffID lookup. This might be an ignorant nonsense, but it seems to me there is _some_ chance that changing how reuse happens would impact how this commit part is done. These local comments on `commitLayer` are not an indication that I understand the big picture, or have a strong opinion about it, right now..)

Reply:
This was fixed here by returning a `stopQueue` boolean.